### PR TITLE
Mirror and detachment scripts

### PR DESCRIPTION
**Pull request by RebeccaStankus**

### CODE REVIEW

File: Shared/Detachment/detachAvatarEntity.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    print("Starting detachment script...............");
+
+    var _entityID, entityPosition, parentJointIndex, parentPosition, checkDetach;
+
+    // get id of reflection area and mirror
+    this.preload = function(entityID) {
+        _entityID = entityID;
+        entityPosition = Entities.getEntityProperties(_entityID, 'position').position;
+        parentJointIndex = Entities.getEntityProperties(_entityID, 'parentJointIndex').parentJointIndex;
+        // print("Attachment ID is : " + _entityID);
```
Comment (misslivirose) [inline Shared/Detachment/detachAvatarEntity.js]: Remove commented out code

File: Shared/Detachment/detachAvatarEntity.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    };
+
+    // distanceToParent = Vec3.distance(entityPosition, parentPosition);
+    // print("Distance to parent joint is : " + distanceToParent);
+    
+    checkDetach = Script.setInterval(function() {
+        entityPosition = Entities.getEntityProperties(_entityID, 'position').position;
+        var distanceToParent = Vec3.distance(entityPosition, parentPosition);
+        print("Distance to parent joint is : " + distanceToParent);
+        if (distanceToParent > 0.6) {
```
Comment (misslivirose) [inline Shared/Detachment/detachAvatarEntity.js]: No magic numbers, give 0.6 a named constant

File: Shared/Detachment/detachAvatarEntity.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+            print("I am no longer in setInterval...goodbye");
```
Comment (misslivirose) [inline Shared/Detachment/detachAvatarEntity.js]: remove print statements that aren't needed

**Comment by RebeccaStankus** [comment]
TEST PLAN FOR AVATAR ENTITY DETACHMENT

-Run this script first to create a watch that is attached to your avatar's left elbow (because hand attachments are not working correctly). You may need to use a default avatar so the watch does not          become hidden inside your avatar.
https://hifi-content.s3.amazonaws.com/rebecca/Detachment/CreateSimpleWatchAttachment.js
-Grab the watch and move it away from the elbow joint.  It should detach when distance from joint is above 0.6.  At this point the watch should no longer move when the left arm moves.

**Comment by RebeccaStankus** [comment]
TEST PLAN FOR MIRROR SCRIPTS
-Run this script to create a mirror in front of your avatar.
https://hifi-content.s3.amazonaws.com/rebecca/spawnMirror.js
-Walk toward the gray side of the mirror (not the side that says "Magico Mirror") and your avatar's reflection should appear.
- To delete the mirror, go to Create>List> and delete "mirror" from there.

**Comment by Nex-Pro** [comment]
Testing this

**Comment by Nex-Pro** [comment]
TEST PLAN FOR AVATAR ENTITY DETACHMENT **PASS**
Needs code review.
Not really part of this PR but watch is oriented to world not the the avatar when attached.

Not sure I'd want the grabbed entity to leave the grabbing hand until it is where I want it to be.

TEST PLAN FOR MIRROR SCRIPTS **PASS**
***About time, it's more than two years since I asked for this!***

Although could not delete as the mirror is locked ( line 68 of the script ) Was unable to clear the lock as, despite having lock perms in my own domain, there seems to be another issue with edit.js in the latest dev build.

**FEEDBACK**
Resolution of the mirror is poor.

Shiny mirror surface washes out the colours of the reflection. This should be tested with a bunch of skyboxes.

Did crash when repeatedly restarted the watch script.

**Comment by misslivirose** [comment]
Thanks @Nex-Pro. The watch itself is just a temporary test and we have a separate PR up for the adjustment of the entities properly. I'll look into what we can do about mirror resolution, as it's using the spectator camera code and may have some restrictions.

**Comment by Nex-Pro** [comment]
@misslivirose No code reviews ?

**Comment by misslivirose** [comment]
Misunderstood about CR on watch, thanks for surfacing

File: Shared/Detachment/detachAvatarEntity.js [hunk 4]
```diff
@@ -22,15 +22,35 @@
...
+            var detachDistance = 0.6;
+            if (distanceToParent > detachDistance) {
+                Entities.editEntity(_entityID, {parentID: "{00000000-0000-0000-0000-000000000000}"});
+                Script.clearInterval(checkDetach);
+            } 
+        }, 100);
+    };
+
+    this.releaseGrab = function(){
+        if (checkDetach){
```
Comment (misslivirose) [inline Shared/Detachment/detachAvatarEntity.js]: space before {

File: Shared/Detachment/detachAvatarEntity.js [hunk 5]
```diff
@@ -22,15 +22,35 @@
...
+                Script.clearInterval(checkDetach);
+            } catch (e) {
+                print("Error stopping interval.");
+            }
+        }
+    };
     
-// exit script
+    this.unload = function(){
+        if (checkDetach){
```
Comment (misslivirose) [inline Shared/Detachment/detachAvatarEntity.js]: space before {

File: Shared/Mirror/mirrorClient.js [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+    var resolution = 1024;	// The resolution of the mirror when turned on
+    var mirrorOverlayID;	// The entity ID of the overlay that displays the mirror reflection
+    var mirrorOverlayRunning;	// True if mirror overlay is reflecting, false otherwise
+    var mirrorOverlayOffset = 0.01;	// The distance between the center of the mirror and the mirror overlay
+    var zeroRot = { w: 1, x: 0, y: 0, z: 0 };	// Constant quaternion for a rotation of 0
+    var mirrorToggleOverlayID;	// The entity ID of the mirror toggle overlay
+    var intervalID;	// The ID of the interval timer used for clearing in unload
+    var editorToClientChannel;	// The channel where mirrorScaler.js sends messages to mirrorClient.js
+    var clientToServerChannel;	// The channel where mirrorClient.js sends messages to mirrorServer.js
+    var serverToClientChannel;	// The channel where mirrorServer.js sends messages to mirrorClient.js
```
Comment (thoys) [inline Shared/Mirror/mirrorClient.js]: serverToClientChannel and editorToClientChannel don't seem to be used, only registered? or am I seeing this wrong?

**Comment by misslivirose** [comment]
@RebeccaStankus - Let's take the detachment script out of this PR given that we've got it integrated into the script in  #11

File: Shared/Mirror/mirrorClient.js (hunk 6 above)
Comment (misslivirose) [reply Shared/Mirror/mirrorClient.js]: @thoys probably remnants of when the rescale handles were attached. @RebeccaStankus let's remove those lines

**Pull request merged**